src/Dependencies/Collections/SegmentedArray.cs

```diff
@@ -0,0 +1,365 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Collections.Internal;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    internal static class SegmentedArray
+    {
+        /// <seealso cref="Array.Clear(Array, int, int)"/>
+        internal static void Clear<T>(SegmentedArray<T> array, int index, int length)
+        {
+            foreach (var memory in array.GetSegments(index, length))
+            {
+                memory.Span.Clear();
+            }
+        }
+
+        /// <seealso cref="Array.Copy(Array, Array, int)"/>
+        internal static void Copy<T>(SegmentedArray<T> sourceArray, SegmentedArray<T> destinationArray, int length)
+        {
+            if (length == 0)
+                return;
+
+            if ((uint)length <= (uint)sourceArray.Length
+                && (uint)length <= (uint)destinationArray.Length)
+            {
+                foreach (var (first, second) in GetSegments(sourceArray, destinationArray, length))
+                {
+                    first.CopyTo(second);
+                }
+
+                return;
+            }
+
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+            if (length > sourceArray.Length)
+                throw new ArgumentException(SR.Arg_LongerThanSrcArray, nameof(sourceArray));
+            if (length > destinationArray.Length)
+                throw new ArgumentException(SR.Arg_LongerThanDestArray, nameof(destinationArray));
+
+            throw new InvalidOperationException("This program location is thought to be unreachable.");
+        }
+
+        private static SegmentEnumerable<T> GetSegments<T>(this SegmentedArray<T> array, int offset, int length)
+            => new(array, offset, length);
+
+        private static AlignedSegmentEnumerable<T> GetSegments<T>(SegmentedArray<T> first, SegmentedArray<T> second, int length)
+            => new(first, second, length);
+
+#pragma warning disable IDE0051 // Remove unused private members (will be used in follow-up)
+        private static AlignedSegmentEnumerable<T> GetSegmentsAligned<T>(SegmentedArray<T> first, int firstOffset, SegmentedArray<T> second, int secondOffset, int length)
+            => new(first, firstOffset, second, secondOffset, length);
+
+        private static UnalignedSegmentEnumerable<T> GetSegmentsUnaligned<T>(SegmentedArray<T> first, int firstOffset, SegmentedArray<T> second, int secondOffset, int length)
+            => new(first, firstOffset, second, secondOffset, length);
+#pragma warning restore IDE0051 // Remove unused private members
+
+        private readonly struct AlignedSegmentEnumerable<T> : IEnumerable<(Memory<T> first, Memory<T> second)>
```

Comment:
>IEnumerable<(Memory<T> first, Memory<T> second)> [](start = 62, length = 48)

Same question about unnecessary implementations of `IEnumerable`/`IEnumerator`

Reply:
➡️ Covered by https://github.com/dotnet/roslyn/pull/50156#discussion_r549876314